### PR TITLE
`contribute` command refactoring

### PR DESCRIPTION
**Pull request by 0xjei**
nb. This activity is mandatory for testing.

- [x] Simplify the workflow by dividing into specific function blocks.
- [x] Move and adapt some core methods (agnostic or specific but reusable) into the `actions` package.
- [x] Allow related cloud functions to return an error (not only logs).
- [x] Optimize contribute pre-conditions + core
- [x] Better error messages.

### CODE REVIEW

File: packages/phase2cli/src/lib/errors.ts [hunk 1]
```diff
@@ -28,9 +29,18 @@ export const COMMAND_ERRORS = {
     COMMAND_ABORT_SELECTION: `The data selection process was suddenly interrupted. Your previous data has not been saved. We are sorry, you will have to repeat the process again from the beginning.`,
     COMMAND_SETUP_NO_R1CS: `Unable to retrieve R1CS files from current working directory. Please, run this command from a working directory where the R1CS files are located to continue with the setup process. We kindly ask you to run the command from an empty directory containing only the R1CS files.`,
     COMMAND_SETUP_DOWNLOAD_PTAU: `Unable to download Powers of Tau file from Hermez Cryptography Phase 1 Trusted Setup. Possible causes may involve an error while making the request (be sure to have a stable internet connection). Please, we kindly ask you to terminate the current session and repeat the process.`,
-    COMMAND_SETUP_ABORT: `You chose to abort the setup process.`
+    COMMAND_SETUP_ABORT: `You chose to abort the setup process.`,
+    COMMAND_CONTRIBUTE_NO_OPENED_CEREMONIES: `Unfortunately, there is no ceremony for which you can make the contribution at this time. Please, try again later.`,
```
Comment (ctrlc03) [inline packages/phase2cli/src/lib/errors.ts:33]: for which you can make **a** contribution

File: packages/actions/test/unit/storage.test.ts [hunk 2]
```diff
@@ -132,43 +128,42 @@ describe("Storage", () => {
...
+                    streamChunkSizeInMb
                 )
                 expect(success).to.be.true
             })
             it("should return true if the object exists", async () => {
                 // login as coordinator
                 await signInWithEmailAndPassword(userAuth, users[1].data.email, passwords[1])
                 // check existence
-                const exists = await objectExist(userFunctions, bucketName, objectName)
+                const exists = await checkIfObjectExist(userFunctions, bucketName, objectName)
```
Comment (ctrlc03) [inline packages/actions/test/unit/storage.test.ts:139]: let's change in describe("") too

File: packages/actions/test/unit/database.test.ts [hunk 3]
```diff
@@ -194,7 +194,7 @@ describe("Database", () => {
 
     describe("getCurrentContributorContribution", () => {
         it("should return an empty array when a ceremony has not participants", async () => {
-            const contributions = await getCurrentContributorContribution(
+            const contributions = await getCircuitContributionsFromContributor(
```
Comment (ctrlc03) [inline packages/actions/test/unit/database.test.ts:197]: let's change it inside describe("") too

File: packages/backend/src/functions/ceremony.ts [hunk 4]
```diff
@@ -55,3 +75,260 @@ export const stopCeremony = functions.pubsub.schedule(`every 30 minutes`).onRun(
         })
     }
 })
+
+/**
+ * Register all ceremony setup-related documents on the Firestore database.
+ * @dev this function will create a new document in the `ceremonies` collection and as needed `circuit`
+ * documents in the sub-collection.
+ */
+export const setupCeremony = functions.https.onCall(
```
Comment (ctrlc03) [inline packages/backend/src/functions/ceremony.ts:84]: should we rename this to setupPhase2Ceremony? given we are setting the properties in Firestore as Phase2

File: packages/backend/src/functions/timeout.ts [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    queryOpenedCeremonies
+} from "../lib/utils"
+import { COMMON_ERRORS, logAndThrowError, printLog, SPECIFIC_ERRORS } from "../lib/errors"
+import { LogLevel } from "../../types/enums"
+
+dotenv.config()
+
+/**
+ * Check and remove the current contributor if it doesn't complete the contribution on the specified amount of time.
+ * @dev since this cloud function is executed every minute, delay problems may occur. See problem #192 for details.
```
Comment (ctrlc03) [inline packages/backend/src/functions/timeout.ts:24]: see issue #192  
maybe let's add a link?

File: packages/backend/src/lib/errors.ts [hunk 6]
```diff
@@ -94,6 +94,42 @@ export const SPECIFIC_ERRORS = {
         "failed-precondition",
         "Unable to interact with a multi-part upload (start, create pre-signed urls or complete).",
         "Authenticated user is not a current contributor which is currently in the uploading step."
+    ),
+    SE_CONTRIBUTE_NO_CEREMONY_CIRCUITS: makeError(
+        "not-found",
+        "There is no circuit associated with the ceremony.",
+        "No documents in the circuits subcollection were found for the selected ceremony."
+    ),
+    SE_CONTRIBUTE_NO_OPENED_CEREMONIES: makeError("not-found", "There are no ceremonies open to contributions."),
```
Comment (ctrlc03) [inline packages/backend/src/lib/errors.ts:103]: I think this is missing one of the 3 descriptions

File: packages/phase2cli/src/lib/prompts.ts [hunk 7]
```diff
@@ -508,96 +514,94 @@ export const promptZkeyGeneration = async (): Promise<boolean> => {
...
         })
     }
 
-    // Ask for selection.
+    // Prompt for selection.
     const { ceremony } = await prompts({
         type: "select",
         name: "ceremony",
-        message: theme.text.bold("Select a ceremony"),
+        message: theme.text.bold("What ceremony do you want to contribute to?"),
```
Comment (ctrlc03) [inline packages/phase2cli/src/lib/prompts.ts:546]: Which ceremony would you like to contribute to

File: packages/phase2cli/src/commands/contribute.ts [hunk 8]
```diff
@@ -6,455 +6,1078 @@ import {
...
+    // nb. this listener encapsulates the core business logic of the contribute command.
+    // the `changedParticipant` is the updated version (w/ newest changes) of the participant's document.
+    const unsubscribe = onSnapshot(participant.ref, async (changedParticipant: DocumentSnapshot) => {
+        // Check data.
+        if (!participant.data() || !changedParticipant.data())
+            showError(COMMAND_ERRORS.COMMAND_CONTRIBUTE_NO_PARTICIPANT_DATA, true)
+
+        // Extract data.
+        const {
+            contributionProgress: exContributionProgress,
```
Comment (ctrlc03) [inline packages/phase2cli/src/commands/contribute.ts:650]: maybe "prev"X is more appropriate?

File: packages/actions/test/unit/storage.test.ts [hunk 9]
```diff
@@ -212,8 +200,7 @@ describe("Storage", () => {
                     bucketName,
                     objectName,
                     localPath,
-                    streamChunkSizeInMb.toString(),
-                    presignedUrlExpirationInSeconds
+                    streamChunkSizeInMb
                 )
                 expect(success).to.be.true
```
Comment (ctrlc03) [inline packages/actions/test/unit/storage.test.ts:205]: These tests need changing due to multiPartUpload not returning true/false anymore. Please change to instead:

assert.isFulfilled(multiPartUpload(...))

File: packages/actions/test/unit/storage.test.ts [hunk 10]
```diff
@@ -223,8 +210,7 @@ describe("Storage", () => {
...
```
Comment (ctrlc03) [inline packages/actions/test/unit/storage.test.ts:215]: as above, change the success criteria to being fulfilled and not true

File: packages/actions/test/unit/storage.test.ts [hunk 11]
```diff
@@ -283,8 +255,7 @@ describe("Storage", () => {
                     bucketName,
                     objectName,
                     localPath,
-                    streamChunkSizeInMb.toString(),
-                    presignedUrlExpirationInSeconds
+                    streamChunkSizeInMb
```
Comment (ctrlc03) [inline packages/actions/test/unit/storage.test.ts:258]: as above, change the success criteria to being fulfilled and not true

File: packages/actions/test/unit/storage.test.ts [hunk 12]
```diff
@@ -132,43 +128,42 @@ describe("Storage", () => {
...
                 )
                 expect(success).to.be.true
```
Comment (ctrlc03) [inline packages/actions/test/unit/storage.test.ts:133]: let's remove the check to be true, and instead check that the promise is fulfilled without throwing

File: packages/backend/src/functions/ceremony.ts [hunk 13]
```diff
@@ -55,3 +75,262 @@ export const stopCeremony = functions.pubsub.schedule(`every 30 minutes`).onRun(
...
+
+/// @todo needs refactoring below.
+
+/**
+ * Add Verifier smart contract and verification key files metadata to the last final contribution for verifiability/integrity of the ceremony.
+ */
+export const finalizeLastContribution = functions.https.onCall(
+    async (data: any, context: functions.https.CallableContext): Promise<any> => {
+        if (!context.auth || !context.auth.token.coordinator)
+            printLog(COMMON_ERRORS.GENERR_NO_COORDINATOR, LogLevel.ERROR)
```
Comment (ctrlc03) [inline packages/backend/src/functions/ceremony.ts:188]: I see in other functions you use directly:

`logAndThrowError(COMMON_ERRORS.CM_NOT_COORDINATOR_ROLE)`

File: packages/phase2cli/src/lib/utils.ts [hunk 14]
```diff
@@ -326,199 +192,54 @@ export const terminate = async (ghUsername: string) => {
...
+    content: string,
+    ceremonyTitle: string,
+    ceremonyPrefix: string
+): Promise<string> => {
+    // Make request.
+    const response = await request("POST /gists", {
+        description: `Attestation for ${ceremonyTitle} MPC Phase 2 Trusted Setup ceremony`,
+        public: true,
+        files: {
+            [`${ceremonyPrefix}_attestation.log`]: {
```
Comment (ctrlc03) [inline packages/phase2cli/src/lib/utils.ts:214]: use `foldersAndPathsTerms.attestation`

File: packages/phase2cli/src/lib/utils.ts [hunk 15]
```diff
@@ -535,789 +256,481 @@ export const getTranscriptLogger = (transcriptFilename: string): Logger =>
...
 
-    // Write first message.
+    // Compute zkey indexes.
+    const lastZkeyIndex = formatZkeyIndex(completedContributions)
+    const nextZkeyIndex = formatZkeyIndex(completedContributions + 1)
+
+    // Prepare zKey filenames.
+    const lastZkeyCompleteFilename = `${circuitPrefix}_${lastZkeyIndex}.zkey`
+    const nextZkeyCompleteFilename = isFinalizing
+        ? `${circuitPrefix}_final.zkey`
```
Comment (ctrlc03) [inline packages/phase2cli/src/lib/utils.ts:496]: use constants "final" var?

File: packages/phase2cli/src/lib/utils.ts [hunk 16]
```diff
@@ -535,789 +256,481 @@ export const getTranscriptLogger = (transcriptFilename: string): Logger =>
...
+    const lastZkeyLocalFilePath = isFinalizing
+        ? getFinalZkeyLocalFilePath(lastZkeyCompleteFilename)
+        : getContributionLocalFilePath(lastZkeyCompleteFilename)
+    const nextZkeyLocalFilePath = isFinalizing
+        ? getFinalZkeyLocalFilePath(nextZkeyCompleteFilename)
+        : getContributionLocalFilePath(nextZkeyCompleteFilename)
+
+    // Generate a custom file logger for contribution transcript.
+    const transcriptCompleteFilename = isFinalizing
+        ? `${circuit.data.prefix}_${contributorOrCoordinatorIdentifier}_final.log`
```
Comment (ctrlc03) [inline packages/phase2cli/src/lib/utils.ts:511]: use constants for final

File: packages/backend/src/functions/circuit.ts [hunk 17]
```diff
@@ -197,12 +207,9 @@ export const coordinateContributors = functionsV1.firestore
             LogLevel.INFO
```
Comment (ctrlc03) [inline packages/backend/src/functions/circuit.ts:207]: I think this log message is missing a word, like "changed"

File: packages/backend/src/functions/participant.ts [hunk 18]
```diff
@@ -0,0 +1,477 @@
...
+
+        // Prepare Firestore DB.
+        const firestore = admin.firestore()
+
+        // Get data.
+        const { ceremonyId } = data
+        const userId = context.auth?.uid
+
+        // Look for the ceremony document.
+        const ceremonyDoc = await getDocumentById(commonTerms.collections.ceremonies.name, ceremonyId)
```
Comment (ctrlc03) [inline packages/backend/src/functions/participant.ts:52]: You usually do a check like: if (!ceremonyDoc.exists) throw - should this be added here?

**Review by ctrlc03 — COMMENTED** [review]
Great stuff, let's a few comments here and there.

File: packages/backend/src/functions/ceremony.ts (hunk 4 above)
Comment (0xjei) [reply packages/backend/src/functions/ceremony.ts:83]: IMHO, I would suggest no for two main reasons: 
1) We are not supporting any Phase 1 right now.
2) When we support the Phase 1, we are going to use a flag to discriminate (e.g., `p0tion setup --phase 1 (or 2)).

Comment (ctrlc03) [reply packages/backend/src/functions/ceremony.ts:83]: sounds good

File: packages/backend/src/lib/errors.ts (hunk 6 above)
Comment (0xjei) [reply packages/backend/src/lib/errors.ts:113]: Yes, the third one (details) is optional. In this particular error, is not needed at all.

File: packages/phase2cli/src/commands/contribute.ts (hunk 8 above)
Comment (0xjei) [reply packages/phase2cli/src/commands/contribute.ts:650]: done

File: packages/backend/src/functions/ceremony.ts (hunk 13 above)
Comment (0xjei) [reply packages/backend/src/functions/ceremony.ts:187]: Yep, this is going to be updated when refactoring the finalize command!

File: packages/phase2cli/src/lib/utils.ts (hunk 14 above)
Comment (0xjei) [reply packages/phase2cli/src/lib/utils.ts:199]: +1 <3

File: packages/backend/src/functions/participant.ts (hunk 18 above)
Comment (0xjei) [reply packages/backend/src/functions/participant.ts:52]: `getDocumentById` itself includes the document existence check!

**Pull request merged**